src/lib/support/CHIPMem.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements parts of the CHIP allocation API that are
+ *      independent of the configured allocator.
+ *
+ */
+
+#include <core/CHIPConfig.h>
+#include <support/CHIPMem.h>
+#include <support/CHIPPlatformMemory.h>
+
+#include <atomic>
+#include <stdlib.h>
+
+namespace chip {
+namespace Platform {
+
+extern CHIP_ERROR MemoryAllocatorInit(void * buf, size_t bufSize);
+extern void MemoryAllocatorShutdown();
+
+static std::atomic_int memoryInitializationCount{ 0 };
+
+CHIP_ERROR MemoryInit(void * buf, size_t bufSize)
```

Comment:
So if someone inits with a null buf from somewhere like TLVReader and then someone else tries to init with a non-null buf, we'll ignore the second call and not actually have a memory area to allocate from?

None of this matters for the malloc-based implementation, of course, but I'm trying to figure out how this will work in practice in the environments that expect to use `CHIPMem-SimpleAlloc`...

Reply:
That is true. It's not unreasonable to think that MemoryInit could add buffers. It's also not unreasonable to think that MemoryShutdown would free anything remaining from the matching MemoryInit. But neither implementation works that way.

We could return an error if the non-initial Init passes a buffer.

(I'm still not convinced that any environment will use `CHIPMem-SimpleAlloc` in preference to a platform-level `malloc` with similar functionality.)

Reply:
Yeah, I find the lack of concrete proposed uses for `CHIPMem-SimpleAlloc` to be a hindrance in terms of designing this API... :(